src/rust/engine/fs/src/snapshot.rs

```diff
@@ -392,37 +384,71 @@ fn osstring_as_utf8(path: OsString) -> Result<String, String> {
     .map_err(|p| format!("{:?}'s file_name is not representable in UTF8", p))
 }
 
+// StoreFileByDigest allows a File to be saved to an underlying Store, in such a way that it can be
+// looked up by the Digest produced by the store_by_digest method.
+// It is a separate trait so that caching implementations can be written which wrap the Store (used
+// to store the bytes) and VFS (used to read the files off disk if needed).
+pub trait StoreFileByDigest<Error> {
+  fn store_by_digest(&self, file: File) -> BoxFuture<Digest, Error>;
```

Comment:
Why change the signature here? Ie pass in a file vs the reference to the file. Is it required because this is only an interface?

Reply:
This is a "my personal taste thing", but I think a reasonable guideline...

tl;dr: Be explicit about overhead. It should be reasonable to expect that if a method takes a reference, it's not going to allocate memory for its own copy of that reference.

Both implementations of this trait actually need to own a `File`; the `OneOffStoreFileByDigest` implementation because it passes it to a thread on a threadpool (so needs a copy, because the calling stack may no longer exist when it's run), and the `Context` implementation because it creates a `DigestFile` which itself needs to own a `File`.

I have a general principle that if you're going to end up taking ownership of something's memory, it's better to be up-front about that and say so in the method signature, than to pretend you just need a reference, and quietly make a copy of it your implementation. It means that when you write a call to the method, you're aware that you're making a copy, and if you think this is expensive (e.g. it's being done in a loop on a large Vec) you can decide whether this is a good idea, or whether it makes sense to e.g. rewrite some code so that you can wrap your type in an `Rc` so that copying it is cheap.

Here, I'm sure it doesn't matter at all, because `File` is tiny, and copying it is cheap, but I personally find the take-a-reference-and-copy-it thing un-idiomatic :)